Framework/Algorithms/src/FFTPreProcessing.cpp

```diff
@@ -0,0 +1,228 @@
+  declareProperty(
+      "ApodizationFunction", "None",
+      boost::make_shared<Mantid::Kernel::StringListValidator>(
+          std::vector<std::string>{"None", "Bartlett", "Lorentz", "Connes",
+                                   "Cosine", "Gaussian", "Welch"}),
+      "The apodization function to apply to the data");
+  declareProperty("DecayConstant", 1.5,
+                  "The decay constant for the apodization function.");
+  auto mustBePositive = boost::make_shared<Kernel::BoundedValidator<int>>();
+  mustBePositive->setLower(0);
+  declareProperty(
+      "Padding", 0, mustBePositive->clone(),
+      "The amount of padding to add to the data,"
+      "it is the number of multiples of the data set."
+      "i.e 0 means no padding and 1 will double the number of data points.");
+}
+
+/** Executes the algorithm
+ *
+ */
+void FFTPreProcessing::exec() {
+  std::vector<int> spectra = getProperty("Spectra");
+
+  // Get original workspace
+  API::MatrixWorkspace_const_sptr inputWS = getProperty("InputWorkspace");
+  auto numSpectra = inputWS->getNumberHistograms();
+  // Create output workspace with same dimensions as input
+  API::MatrixWorkspace_sptr outputWS = getProperty("OutputWorkspace");
+  if (inputWS != outputWS) {
+    outputWS = API::WorkspaceFactory::Instance().create(inputWS);
+  }
+
+  // Share the X values
+  for (size_t i = 0; i < static_cast<size_t>(numSpectra); ++i) {
+    outputWS->setSharedX(i, inputWS->sharedX(i));
+  }
+
+  // No spectra specified = process all spectra
+  if (spectra.empty()) {
+    spectra = std::vector<int>(numSpectra);
+    std::iota(spectra.begin(), spectra.end(), 0);
+  }
+
+  Progress prog(this, 0.0, 1.0, numSpectra + spectra.size());
+  if (inputWS != outputWS) {
+
+    // Copy all the Y and E data
+    PARALLEL_FOR_IF(Kernel::threadSafe(*inputWS, *outputWS))
+    for (int64_t i = 0; i < int64_t(numSpectra); ++i) {
+      PARALLEL_START_INTERUPT_REGION
+      const auto index = static_cast<size_t>(i);
+      outputWS->setSharedY(index, inputWS->sharedY(index));
+      outputWS->setSharedE(index, inputWS->sharedE(index));
+      prog.report();
+      PARALLEL_END_INTERUPT_REGION
+    }
+    PARALLEL_CHECK_INTERUPT_REGION
+  }
+  const std::string method = getProperty("ApodizationFunction");
+  const double decayConstant = getProperty("DecayConstant");
+  const int padding = getProperty("Padding");
+  fptr apodizationFunction = getApodizationFunction(method);
+  // Do the specified spectra only
+  int specLength = static_cast<int>(spectra.size());
+  std::vector<double> norm(specLength, 0.0);
+  PARALLEL_FOR_IF(Kernel::threadSafe(*inputWS, *outputWS))
+  for (int i = 0; i < specLength; ++i) {
+    PARALLEL_START_INTERUPT_REGION
+    const auto specNum = static_cast<size_t>(spectra[i]);
+
+    if (spectra[i] > static_cast<int>(numSpectra)) {
+      g_log.error("The spectral index " + std::to_string(spectra[i]) +
+                  " is greater than the number of spectra!");
```

Comment:
When you throw from `exec` `Algorithm::execute` catches it and logs. Simply delete the lines.